### PR TITLE
Expressive and Zf3 support

### PR DESCRIPTION
**Pull request by wshafer**
I know we are talking about this in another PR.  However, I needed to do this anyway for another project on Expressive.

This PR essentially adds a middleware layer for expressive.  Updates the factories for Service Manager 3, removes dependencies of ZF2, adds config for the new zend-component-installer and updates the tests so they'll run in php7.

There's still some things that can be improved on, but I thought I'd submit this for review, discussion, and perhaps a good jumping off point.

### CODE REVIEW

**Comment by andy-1bg** [comment]
+1

**Comment by RWOverdijk** [comment]
@wshafer The other pr has been merged.

**Comment by wshafer** [comment]
@RWOverdijk - All changes merged.

**Comment by RWOverdijk** [comment]
Oof huge diff. I'll take a look at this at a more quiet time.

**Comment by wshafer** [comment]
@RWOverdijk - Thanks.   If you like I can break this up into a couple of smaller PR's that do just ONE thing each.  But perhaps a new branch to PR against might be helpful, until it's all ready?

**Comment by RWOverdijk** [comment]
@wshafer So far (~40% in) the code looks good. I must admit I haven't done anything with expressive so it's difficult for me to judge. I'd like someone who has worked with it to give a second opinion on this PR.

**Comment by RWOverdijk** [comment]
So, coming this PR, we'll no longer support zf2. Still digging through it. Sorry it takes so long.

File: composer.json [hunk 1]
```diff
@@ -1,13 +1,14 @@
...
-    "description": "An assetmanager module for Zend Framework 2.",
+    "description": "An assetmanager module for Zend Framework 3 and Zend Expressive.",
     "license": "BSD-2-Clause",
     "type": "library",
     "keywords": [
         "asset",
         "asset manager",
         "zf2",
-        "zend framework 2"
+        "zend framework 2",
```
Comment (RWOverdijk) [inline composer.json]: 3

File: composer.json [hunk 2]
```diff
@@ -1,13 +1,14 @@
 {
     "name": "rwoverdijk/assetmanager",
-    "description": "An assetmanager module for Zend Framework 2.",
+    "description": "An assetmanager module for Zend Framework 3 and Zend Expressive.",
     "license": "BSD-2-Clause",
     "type": "library",
     "keywords": [
         "asset",
         "asset manager",
         "zf2",
```
Comment (RWOverdijk) [inline composer.json]: 3

File: composer.json [hunk 3]
```diff
@@ -23,25 +24,19 @@
...
     "require": {
-        "php": "^5.5 || ^7.0",
-        "zendframework/zend-modulemanager": "^2.7.2",
-        "zendframework/zend-stdlib": "^2.7.7 || ^3.0.1",
-        "zendframework/zend-servicemanager": "^2.7.6 || ^3.1.1",
-        "zendframework/zend-loader": "^2.5.1",
-        "zendframework/zend-eventmanager": "^2.6.3 || ^3.0.1",
-        "zendframework/zend-mvc": "^2.7.9 || ^3.0.2",
-        "kriswallsmith/assetic": "^1.2.1"
+        "php": ">=5.5",
```
Comment (RWOverdijk) [inline composer.json]: 7

File: config/expressive.config.php [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+            \AssetManager\MiddleWare\AssetManagerMiddleware::class => \AssetManager\MiddleWare\AssetManagerMiddlewareFactory::class,
+        ),
+        'invokables' => array(
+            'AssetManager\Service\MimeResolver'                 => 'AssetManager\Service\MimeResolver',
+        ),
+    ),
+    'asset_manager' => array(
+        'clear_output_buffer' => true,
+        'resolvers' => array(
+            'AssetManager\Resolver\MapResolver'                 => 3000,
```
Comment (RWOverdijk) [inline config/expressive.config.php]: Align

File: config/expressive.config.php [hunk 5]
```diff
@@ -0,0 +1,57 @@
...
+            'AssetManager\Resolver\AliasPathStackResolver'      => 'AssetManager\Service\AliasPathStackResolverServiceFactory',
+            \AssetManager\MiddleWare\AssetManagerMiddleware::class => \AssetManager\MiddleWare\AssetManagerMiddlewareFactory::class,
+        ),
+        'invokables' => array(
+            'AssetManager\Service\MimeResolver'                 => 'AssetManager\Service\MimeResolver',
+        ),
+    ),
+    'asset_manager' => array(
+        'clear_output_buffer' => true,
+        'resolvers' => array(
```
Comment (RWOverdijk) [inline config/expressive.config.php]: Align array

File: config/expressive.config.php [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+            'AssetManager\Resolver\MapResolver'                 => 'AssetManager\Service\MapResolverServiceFactory',
+            'AssetManager\Resolver\PathStackResolver'           => 'AssetManager\Service\PathStackResolverServiceFactory',
+            'AssetManager\Resolver\PrioritizedPathsResolver'    => 'AssetManager\Service\PrioritizedPathsResolverServiceFactory',
+            'AssetManager\Resolver\CollectionResolver'          => 'AssetManager\Service\CollectionResolverServiceFactory',
+            'AssetManager\Resolver\ConcatResolver'              => 'AssetManager\Service\ConcatResolverServiceFactory',
+            'AssetManager\Resolver\AliasPathStackResolver'      => 'AssetManager\Service\AliasPathStackResolverServiceFactory',
+            \AssetManager\MiddleWare\AssetManagerMiddleware::class => \AssetManager\MiddleWare\AssetManagerMiddlewareFactory::class,
+        ),
+        'invokables' => array(
+            'AssetManager\Service\MimeResolver'                 => 'AssetManager\Service\MimeResolver',
```
Comment (RWOverdijk) [inline config/expressive.config.php]: Align

File: config/expressive.config.php [hunk 7]
```diff
@@ -0,0 +1,57 @@
+<?php
+return array(
+    'dependencies' => array (
```
Comment (RWOverdijk) [inline config/expressive.config.php]: I think you can update this file. [currently](https://github.com/RWOverdijk/AssetManager/blob/master/config/module.config.php#L5) we use the class names for all of these.

File: config/expressive.config.php [hunk 8]
```diff
@@ -0,0 +1,57 @@
+<?php
```
Comment (RWOverdijk) [inline config/expressive.config.php]: I'm curious why expressive gets its own config file. But something tells me I'll find out further down :). Either way, I think we can re-use a lot of the configuration being copied here.

File: public/test.js [hunk 9]
```diff
@@ -0,0 +1 @@
+alert('Hi');
```
Comment (RWOverdijk) [inline public/test.js]: I think you dropped something

**Comment by RWOverdijk** [comment]
Looking at the code, I think it first needs to get the coding style applied throughout the PR. Maybe nice, too, would be inline comments to explain what's happening.

File: config/expressive.config.php (hunk 8 above)
Comment (wshafer) [inline config/expressive.config.php]: It uses different config keys.  service_manager is replaced by "dependencies", BUT that doesn't mean we can't include the other config file AND re-key it.  That might be a better way to go here.

**Comment by wshafer** [comment]
There's another strange "gotcha" that's been bothering about this PR too.  the PSR7 Bridge is now required for the module to work with ZF3, yet we can't add that as a hard dependency for expressive.

There's a few suggestions...  

1) Make composer "suggest" this package for ZF3 and update the install instructions on the readme file and wiki.
2) Separate the "core" library and create meta-packages for zf3 and expressive.

Option 1 seems best short-term to me.  Thoughts?

File: config/expressive.config.php (hunk 8 above)
Comment (RWOverdijk) [inline config/expressive.config.php]: I think I might agree. The keys are pretty much the same as far as I can see

**Comment by RWOverdijk** [comment]
@wshafer Short term option 1, then create an issue for option 2.

I think splitting this PR up in smaller PRs might be a good idea. It's pretty difficult to review a PR of this size. Especially because I don't know expressive.

**Comment by wshafer** [comment]
Absolutely.  

How's this road map?

1) Get tests running under PHP 7
2) Fix factories for container-interop, ZF3 BC breaks, and Composer updates
3) Add PSR 7 support + PSR7Bridge for ZF3
4) Add Expressive middleware
5) Add Github issue to break out the ZF2 and Expressive modules.

**Comment by wshafer** [comment]
Actually... now that I think about it...  step 4 should come before 5.

**Comment by RWOverdijk** [comment]
That sounds like a fine plan to me. Makes for smaller chunks that can be processed one at a time.

**Comment by wshafer** [comment]
@RWOverdijk - Sorry for the long hiatus, but I'm back and still needing a PSR7 Container-Interop solution for assets.  This is still the best option out there.

Before I start I just want to touch base.   Are we ready for `ZF3 BC breaks, and Composer updates` or creating a core lib repo so we can break out expressive from zf3?

**Comment by RWOverdijk** [comment]
@wshafer I'm fine with BC breaks on a major version bump, absolutely!

**Comment by wshafer** [comment]
Cool  I'll get started.  Thanks

**Pull request closed without merging**